flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/UnionInputGate.java

```diff
@@ -138,6 +148,24 @@ public void requestPartitions() throws IOException, InterruptedException {
 		}
 	}
 
+	@Override
+	public void blockInputChannel(int channelIndex) {
+		InputGate inputGate = indexToInputGateMap.get(channelIndex);
+		if (inputGate == null) {
+			throw new IllegalStateException("Could not find input gate from the channel index " + channelIndex);
+		}
+
+		int indexOffset = inputGateToIndexOffsetMap.get(inputGate);
+		inputGate.blockInputChannel(channelIndex - indexOffset);
```

Comment:
wrap this offset handling in a private method:
```
inputGate.blockInputChannel(externalToInternalChannelIndex(channelIndex, inputGate));
```
and similar call in `getNextBufferOrEvent()` method to:
```
bufferOrEvent.setChannelIndex(internalToExternalChannelIndex(bufferOrEvent.getChannelIndex(), inputGate));
```

Reply:
alright